### PR TITLE
Let coverage for Kotlin DSL TAPI models do TAPI/Gradle cross-version testing

### PR DESCRIPTION
**Pull request by eskatos**
The Kotlin DSL provides TAPI models to supporting IDEs. This PR moves all the existing coverage to cross-version TAPI tests in order to cover the supported Tooling API versions and Gradle versions.

The Kotlin DSL TAPI models aren't bundled into the public `gradle-tooling-api` jar. To support this setup, the `ToolingApiSpecification` was augmented to allow descendant test classes to declaratively provide additional client-side classpath. The Kotlin DSL TAPI cross-version tests use that to provide the test runner with `:kotlinDslToolingModels` and its dependencies (`kotlin-stdlib`) from the Gradle distribution under test. Mimicking the supporting IDEs behaviour.

There are two TAPI models, `KotlinBuildScriptTemplateModel` and `KotlinBuildScriptModel`. All tests require TAPI >= 4.1. The test for `KotlinBuildScriptTemplateModel` targets Gradle >= 4.1. All tests for the more evolved `KotlinBuildScriptModel` required changes in `:kotlinDsl` in order to provide the `GradleConnector` set up by `ToolingApiSpecification` hence target Gradle >= 5.4.

### CODE REVIEW

**Comment by eskatos** [comment]
[all versions tests on linux](https://builds.gradle.org/viewLog.html?buildId=20886222) ✅ 
[all versions tests on windows](https://builds.gradle.org/viewLog.html?buildId=20887853) ✅

File: subprojects/kotlin-dsl-tooling-builders/src/crossVersionTest/groovy/org/gradle/kotlin/dsl/tooling/builders/r54/KotlinBuildScriptModelCrossVersionSpec.groovy [hunk 1]
```diff
@@ -0,0 +1,574 @@
...
+            """ plugins { kotlin("jvm") version "$targetKotlinVersion" } """
+        )
+    }
+
+    @LeaksFileHandles("Kotlin compiler daemon on buildSrc jar")
+    def "sourcePath includes buildSrc source roots"() {
+
+        given:
+        withKotlinBuildSrc()
+        withSettings("""include(":sub")""")
```
Comment (bamboo) [inline subprojects/kotlin-dsl-tooling-builders/src/crossVersionTest/groovy/org/gradle/kotlin/dsl/tooling/builders/r54/KotlinBuildScriptModelCrossVersionSpec.groovy]: :x: Better to augment the default settings like we do in other places, `withDefaultSettings().appendText(...)`

File: subprojects/kotlin-dsl-tooling-builders/src/crossVersionTest/groovy/org/gradle/kotlin/dsl/tooling/builders/r54/PrecompiledScriptPluginModelCrossVersionSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+    @LeaksFileHandles("Kotlin Compiler Daemon working directory")
+    def "given a single project build, the classpath of a precompiled script plugin is the compile classpath of its enclosing source-set"() {
+
+        given:
+        def implementationDependency = withEmptyJar("implementation.jar")
+        def classpathDependency = withEmptyJar("classpath.jar")
+
+        and:
+        withDefaultSettings()
```
Comment (bamboo) [inline subprojects/kotlin-dsl-tooling-builders/src/crossVersionTest/groovy/org/gradle/kotlin/dsl/tooling/builders/r54/PrecompiledScriptPluginModelCrossVersionSpec.groovy]: :x: `withDefaultSettings` was part of the test setup in https://github.com/gradle/gradle/blob/91ed97f81ee638a24915bb20a94db1feb3058eb6/subprojects/kotlin-dsl-test-fixtures/src/main/kotlin/org/gradle/kotlin/dsl/fixtures/AbstractPluginTest.kt#L19

File: subprojects/kotlin-dsl-tooling-builders/src/crossVersionTest/groovy/org/gradle/kotlin/dsl/tooling/builders/r54/KotlinBuildScriptModelCrossVersionSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,574 @@
...
+                    classpath(files("classes.jar"))
+                }
+            }
+
+            val p =
+        """.stripIndent()
+
+        expect:
+        assertClassPathContains(
+            file("classes.jar"))
```
Comment (bamboo) [inline subprojects/kotlin-dsl-tooling-builders/src/crossVersionTest/groovy/org/gradle/kotlin/dsl/tooling/builders/r54/KotlinBuildScriptModelCrossVersionSpec.groovy]: ⭕️ Let's take the opportunity to close all multiline closing brackets on a separate line, i.e.

```
assertClassPathContains(
    file("classes.jar")
)
```

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/resolver/KotlinBuildScriptModelRequest.kt [hunk 4]
```diff
@@ -69,67 +70,120 @@ typealias ModelBuilderCustomization = ModelBuilder<KotlinBuildScriptModel>.() ->
...
-    val effectiveProjectDir = buildSrcProjectDirOf(scriptFile, importedProjectDir)
-        ?: importedProjectDir
 
-    val scriptModel = fetchKotlinBuildScriptModelFrom(effectiveProjectDir, request, modelBuilderCustomization)
-    if (scriptModel.enclosingScriptProjectDir == null && hasProjectDependentClassPath(scriptFile)) {
-        val externalProjectRoot = projectRootOf(scriptFile, importedProjectDir)
-        if (externalProjectRoot != importedProjectDir) {
-            return fetchKotlinBuildScriptModelFrom(externalProjectRoot, request, modelBuilderCustomization)
+private
+data class FetchParameters(
```
Comment (bamboo) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/resolver/KotlinBuildScriptModelRequest.kt]: 👍

File: subprojects/kotlin-dsl-tooling-builders/src/crossVersionTest/groovy/org/gradle/kotlin/dsl/tooling/builders/r54/PrecompiledScriptPluginModelCrossVersionSpec.groovy (hunk 2 above)
Comment (bamboo) [reply subprojects/kotlin-dsl-tooling-builders/src/crossVersionTest/groovy/org/gradle/kotlin/dsl/tooling/builders/r54/PrecompiledScriptPluginModelCrossVersionSpec.groovy]: On a second though, now that most logic has been moved to `kotlin-dsl-provider-plugins` it shouldn't be a problem. Better to get coverage of all published versions.

File: subprojects/tooling-api/src/testFixtures/groovy/org/gradle/integtests/tooling/fixture/ToolingApiCompatibilitySuiteRunner.groovy [hunk 5]
```diff
@@ -54,7 +56,9 @@ class ToolingApiCompatibilitySuiteRunner extends AbstractCompatibilityTestRunner
 
         def distribution = versionedTool.distribution
         if (distribution.toolingApiSupported) {
+            // current vs. target
```
Comment (bamboo) [inline subprojects/tooling-api/src/testFixtures/groovy/org/gradle/integtests/tooling/fixture/ToolingApiCompatibilitySuiteRunner.groovy]: 👍

File: subprojects/kotlin-dsl-tooling-builders/src/crossVersionTest/groovy/org/gradle/kotlin/dsl/tooling/builders/AbstractKotlinScriptModelCrossVersionTest.groovy [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+        repositories {
+            gradlePluginPortal()
+        }
+    """.stripIndent()
+
+    private String targetKotlinVersion
+
+    protected String getTargetKotlinVersion() {
+        if (targetKotlinVersion == null) {
+            def props = new JarTestFixture(targetDist.gradleHomeDir.file("lib").listFiles().find {
```
Comment (bamboo) [inline subprojects/kotlin-dsl-tooling-builders/src/crossVersionTest/groovy/org/gradle/kotlin/dsl/tooling/builders/AbstractKotlinScriptModelCrossVersionTest.groovy]: :o: I'd extract the `"kotlin"` property reading to its own function for clarity

**Pull request merged**